reference/best_practices/reusable_bundles.rst

```diff
@@ -1,4 +1,164 @@
+
+Try your bundle with other catalogs. Don't forget use cases with many locales, many channels and various catalog volumes.
+You can use our `data generator bundle`_ to generate a catalog with random data.
+
+.. _Akeneo PIM Community Edition: https://github.com/akeneo/pim-community-standard
+.. _Enterprise Edition: https://www.akeneo.com/enterprise-edition/
+.. _data generator bundle: https://github.com/akeneo-labs/DataGeneratorBundle
+
+
+Maintainability
+---------------
+
+Keep in mind the `Single Responsibility principle`_ from the five `SOLID principles`_ of Object Oriented Programming.
+It seems simple but we often forget to explode your code in multiple classes to make it more readable and testable.
+
+You can use `PMD`_ to help make your code simple, flexible and easily testable.
+
+
+In Akeneo PIM, we use `PhpSpec`_ as unit testing tool and `Behat`_ for behavior-oriented tests.
+
+Use a Continuous Integration (CI) tool to deploy your application, test your code and improve its quality.
+On our open-source projects, we use free CI tools `Travis`_ for unit tests and `Scrutinizer`_ for static analysis.
+You can find some examples on our `Akeneo-Labs`_ projects (`scrutinizer.yml`_ and `travis.yml`_).
+
+.. _Single Responsibility: https://en.wikipedia.org/wiki/Single_responsibility_principle
+.. _SOLID principles: https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)
+.. _PMD: https://phpmd.org/
+.. _PhpSpec: http://phpspec.readthedocs.org/
+.. _Behat: http://docs.behat.org/
+.. _Travis: https://travis-ci.org/
+.. _Scrutinizer: https://scrutinizer-ci.com/
+.. _Akeneo-Labs: https://github.com/akeneo-labs
+.. _scrutinizer.yml: https://github.com/akeneo-labs/CustomEntityBundle/blob/master/.scrutinizer.yml
+.. _travis.yml: https://github.com/akeneo-labs/CustomEntityBundle/blob/master/.travis.yml
+
+
+Another important point about maintainability is to rely on interfaces instead of concrete classes.
+It will facilitate you future migrations of Akeneo PIM.
+
+Don't forget you can use composition instead of inheritance.
+Decorating a class will allow you to benefit from inheritance avoiding migration problems.
+Indeed, class constructors can be changed from a minor version to another, while an interface will never change.
+
+To give integrators a clean way to rely on your code, you can also create your own interfaces.
+
+Extensibility
+-------------
+
+Follow Akeneo PIM dependencies
+""""""""""""""""""""""""""""""
+You **MUST** follow Akeneo PIM dependencies at all costs. Don't change version of existing ones.
+You may add new dependencies needed for your customisations.
+
+
+Use registries
+""""""""""""""
+Registries are also a good extension point when you have to deal with many cases (different attribute types, entities, etc.).
+They are used to gather related classes (i.e: filters, sorters). You can add new classes to a registry simply by tagging it.
```

Comment:
Not sure it's obvious for anyone. What do you think of :
"simply declare it as a service and tag it"
Plus a link to : http://symfony.com/doc/2.7/components/dependency_injection/tags.html